### PR TITLE
add(main): new package libpngwriter and libpngwriter-static

### PR DESCRIPTION
**Pull request by 2096779623**

### CODE REVIEW

**Review by Grimler91 — CHANGES_REQUESTED** [review]
The hostbuild step here doesn't seem to do anything, what is the purpose of it?

Looks like only the static .a lib is created, probably you want to enable the .so by adding `-DBUILD_SHARED_LIBS=ON` to `TERMUX_PKG_EXTRA_CONFIGURE_ARGS`

File: packages/libpngwriter/build.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+TERMUX_PKG_HOMEPAGE=http://pngwriter.sourceforge.net
+TERMUX_PKG_DESCRIPTION="C++ library for creating PNG images"
+TERMUX_PKG_LICENSE="GPL-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=0.7.0
+TERMUX_PKG_SRCURL=https://github.com/pngwriter/pngwriter/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=82d46eef109f434f95eba9cf5908710ae4e75f575fd3858178ad06e800152825
+TERMUX_PKG_DEPENDS="zlib,freetype,libpng"
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_FORCE_CMAKE=true
```
Comment (Grimler91) [inline packages/libpngwriter/build.sh:10]: Should be no need to set FORCE_CMAKE either, there are no other config alternatives

**Comment by 2096779623** [comment]
> The hostbuild step here doesn't seem to do anything, what is the purpose of it?
> 
> Looks like only the static .a lib is created, probably you want to enable the .so by adding `-DBUILD_SHARED_LIBS=ON` to `TERMUX_PKG_EXTRA_CONFIGURE_ARGS`

I need static library, but there is no static library after adding `-DBUILD_SHARED_LIBS=ON`, how can I do this?

**Comment by Grimler91** [comment]
Would require either re-configuring and make'ing pngwriter with different options, or patching the cmake file to add another library target, see https://stackoverflow.com/questions/2152077/is-it-possible-to-get-cmake-to-build-both-a-static-and-shared-library-at-the-sam

**Comment by 2096779623** [comment]
> Would require either re-configuring and make'ing pngwriter with different options, or patching the cmake file to add another library target, see https://stackoverflow.com/questions/2152077/is-it-possible-to-get-cmake-to-build-both-a-static-and-shared-library-at-the-sam

Fixed,thanks!

File: packages/libpngwriter/CMakeLists.txt.patch [hunk 2]
```diff
@@ -0,0 +1,32 @@
+diff -uNr pngwriter-0.7.0/CMakeLists.txt pngwriter-0.7.0.mod/CMakeLists.txt
+--- pngwriter-0.7.0/CMakeLists.txt	2018-01-22 21:10:17.000000000 +0800
++++ pngwriter-0.7.0.mod/CMakeLists.txt	2022-05-30 12:47:32.809998828 +0800
+@@ -70,11 +70,19 @@
+ 
+ # Targets #####################################################################
+ #
++
++INCLUDE_DIRECTORIES(/data/data/com.termux/files/usr/include/freetype2)
```
Comment (Grimler91) [inline packages/libpngwriter/CMakeLists.txt.patch:9]: Please replace `/data/data/com.termux/files/usr` with `@TERMUX_PREFIX@` in patches (which is then [automatically sed'ed](https://github.com/termux/termux-packages/blob/master/scripts/build/termux_step_patch_package.sh#L27))

File: packages/libpngwriter/CMakeLists.txt.patch [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+     src/pngwriter.h
+ )
+ 
++add_library(PNGwriter_shared SHARED
++	src/pngwriter.cc
++	src/pngwriter.h
++)
++
+ target_link_libraries(PNGwriter PUBLIC PNG::PNG)
+ target_link_libraries(PNGwriter PUBLIC ZLIB::ZLIB)
```
Comment (Grimler91) [inline packages/libpngwriter/CMakeLists.txt.patch:22]: You likely need to link PNGwriter_shared against PNG, ZLIB and m as well

File: packages/libpngwriter/build.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+TERMUX_PKG_LICENSE="GPL-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=0.7.0
+TERMUX_PKG_SRCURL=https://github.com/pngwriter/pngwriter/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=82d46eef109f434f95eba9cf5908710ae4e75f575fd3858178ad06e800152825
+TERMUX_PKG_DEPENDS="zlib,freetype,libpng"
+TERMUX_PKG_AUTO_UPDATE=true
+
+termux_step_post_make_install() {
+        mv "$TERMUX_PREFIX"/lib/libPNGwriter_shared.so libPNGwriter.so
```
Comment (Grimler91) [inline packages/libpngwriter/build.sh:12]: Is this really needed when you have `install(TARGETS PNGwriter_shared LIBRARY DESTINATION lib)`?

**Review by Grimler91 — CHANGES_REQUESTED** [review]
Thanks, some comments on the patch

File: packages/libpngwriter/CMakeLists.txt.patch (hunk 2 above)
Comment (2096779623) [reply packages/libpngwriter/CMakeLists.txt.patch:9]: I forgot to do this :(

File: packages/libpngwriter/build.sh (hunk 4 above)
Comment (2096779623) [reply packages/libpngwriter/build.sh:12]: Yes, you need to use the correct file name.

**Comment by 2096779623** [comment]
> Thanks, some comments on the patch

completed.

File: packages/libpngwriter/CMakeLists.txt.patch [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
++target_link_libraries(PNGwriter_shared PUBLIC PNG::PNG)
++target_link_libraries(PNGwriter_shared PUBLIC ZLIB::ZLIB)
+ if(NOT WIN32)
+     # automatically added on windows
+     target_link_libraries(PNGwriter PRIVATE m)
```
Comment (Grimler91) [inline packages/libpngwriter/CMakeLists.txt.patch:27]: I would guess that the shared library needs to be linked against libm here as well

**Review by Grimler91 — CHANGES_REQUESTED** [review]

File: packages/libpngwriter/CMakeLists.txt.patch (hunk 5 above)
Comment (2096779623) [reply packages/libpngwriter/CMakeLists.txt.patch:27]: image

**Pull request merged**